toolkit/pallets/block-production-log/src/lib.rs

```diff
@@ -0,0 +1,94 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+pub mod mock;
+#[cfg(test)]
+pub mod test;
+
+mod benchmarking;
+
+pub mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use sp_block_production_log::*;
+	use sp_consensus_slots::Slot;
+	use sp_runtime::BoundedVec;
+	use sp_std::vec::Vec;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		#[pallet::constant]
+		type MaxLogSize: Get<u32>;
+
+		#[pallet::constant]
+		type MaxProducerIdLength: Get<u32>;
+
+		type BlockProducerId: Member
+			+ Parameter
+			+ MaxEncodedLen
+			+ From<BoundedVec<u8, Self::MaxProducerIdLength>>;
+
+		type WeightInfo: WeightInfo;
+
+		fn current_slot() -> Slot;
+	}
+
+	#[pallet::storage]
+	pub type CurrentBlockProducerId<T: Config> = StorageValue<_, T::BlockProducerId, OptionQuery>;
+
+	#[pallet::storage]
+	pub type Log<T: Config> =
+		StorageValue<_, BoundedVec<(Slot, T::BlockProducerId), T::MaxLogSize>, ValueQuery>;
+
+	#[pallet::inherent]
+	impl<T: Config> ProvideInherent for Pallet<T> {
+		type Call = Call<T>;
+		type Error = InherentError;
+		const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+
+		fn create_inherent(data: &InherentData) -> Option<Self::Call> {
+			let id = data
+				.get_data::<T::BlockProducerId>(&Self::INHERENT_IDENTIFIER)
+				.expect("Block Production Log inherent data not correctly encoded")
+				.expect("Block Production Log inherent data must be provided");
+			Some(Call::append { id })
+		}
+
+		fn is_inherent(call: &Self::Call) -> bool {
+			matches!(call, Call::append { .. })
+		}
+
+		fn is_inherent_required(_: &InherentData) -> Result<Option<Self::Error>, Self::Error> {
+			Ok(Some(Self::Error::InherentRequired))
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight((T::WeightInfo::append(), DispatchClass::Mandatory))]
+		pub fn append(origin: OriginFor<T>, id: T::BlockProducerId) -> DispatchResult {
+			ensure_none(origin)?;
+			let slot = T::current_slot();
+			Log::<T>::try_append((slot, id)).map_err(|_| sp_runtime::DispatchError::Exhausted)
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn take_prefix(slot: &Slot) -> Vec<(Slot, T::BlockProducerId)> {
+			let (to_return, to_retain) = Log::<T>::take().into_iter().partition(|(s, _)| s <= slot);
```

Comment:
maybe it is worth to consider storing data ordered, so taking of prefix may be faster:
https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.SortedMembers.html

Reply:
for instance pallet-membership maintains vector and exposes it via this trait 
https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/membership/src/lib.rs#L394

Reply:
They use exactly same structure as I do. 
```
#[pallet::storage]
pub type Members<T: Config<I>, I: 'static = ()> =
     StorageValue<_, BoundedVec<T::AccountId, T::MaxMembers>, ValueQuery>;
```
Having said that I'll ensure that `slot` is greater than the last currently stored.
This will at same time fix the problem with possible multiple calls.

Reply:
Right. I am not sure who will be using `take_prefix`, but maybe it will be worth to change it and expose it via SortedMembers, so end user will be able to make some additional operations on it, without adjusting logic of this pallet.
I am currious if something like `pub fn take_prefix(slot: &Slot) -> impl SortedMembers<..> {` will work

Reply:
If only `append` is use, this is sorted all the time. Appending with slot lower than the latest slot is illegal.

Reply:
Right, but as we discussed, once it is sorted then we can try also something like binary_search instead of partition

Reply:
`partition` is `O(n)` . Even if we could use something like `binary_search` we still have to copy either one or another part or both parts, so it still has same complexity. At best we could optimize for some scenario when assume that part to retain is much smaller, but I think it won't be easy to best this very simple solution with `partition`.

BTW `binary_search` requires exact hit also, as I allowed multiple blocks in same slot, it is hard to get the "right-most" one.